File: WOT-iOS/WOTData/WOTData/CoreData/model/mapping/manual/jsonparsing/Chassis/WOTWebResponseAdapterChassis.m
//
//  WOTWebResponseAdapterChassis.m
//  WOT-iOS
//
//  Created on 6/23/15.
//  Copyright (c) 2015. All rights reserved.
//

#import "WOTWebResponseAdapterChassis.h"
#import "WOTData.h"
#import <WOTPivot/WOTPivot.h>
#import <WOTData/WOTData-Swift.h>

@implementation WOTWebResponseAdapterChassis

- (void)parseJSON:(NSDictionary *)json error:(NSError *)error {

    if (error) {
        
        debugError(@"%@",error.localizedDescription);
        return;
    }
    
    NSDictionary *tankChassesDictionary = json[WGJsonFields.data];
    
    NSArray *tankChassisArray = [tankChassesDictionary allKeys];
    id<WOTCoredataProviderProtocol> dataProvider = [WOTTankCoreDataProvider sharedInstance];
    NSManagedObjectContext *context = [dataProvider workManagedObjectContext];
    [context performBlock:^{
        
        [tankChassisArray enumerateObjectsUsingBlock:^(NSString *key, NSUInteger idx, BOOL *stop) {
            
            NSDictionary *tankChassisJSON = tankChassesDictionary[key];
            if (![tankChassisJSON isKindOfClass:[NSDictionary class]]){
                
                debugError(@"error while parsing");
                return;
            }
            
            NSPredicate *predicate = [NSPredicate predicateWithFormat:@"%K == %@",WGJsonFields.module_id,tankChassisJSON[WGJsonFields.module_id]];
            Tankchassis *tankChasses = (Tankchassis *)[Tankchassis findOrCreateObjectWithPredicate:predicate context:context];
            [tankChasses mappingFrom:tankChassisJSON];
        }];
        
        if ([context hasChanges]) {
            
            NSError *error = nil;
            [context save:&error];
        }
    }];
}


@end

File: WOT-iOS/WOTData/WOTData/CoreData/model/mapping/manual/jsonparsing/Chassis/WOTWebResponseAdapterChassis.swift
//
//  WOTWebResponseAdapterChassis.swift
//  WOTData
//
//  Created by Pavel Yeshchyk on 1/15/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc
public class WOTWebResponseAdapterChassis: NSObject, WOTWebResponseAdapter {
    
    public func parseJSON(_ json: JSON, error err: NSError?, nestedRequestsCallback: JSONMappingCompletion?) {
        if let error = err {
            print("\(#file) \(#function) at \(error.localizedDescription)")
            return
        }
        guard let data = json[WGJsonFields.data] as? [AnyHashable: Any] else {
            return
        }
        
        let keys = data.keys
        guard keys.count != 0 else {
            return
        }

        let context = WOTTankCoreDataProvider.sharedInstance.workManagedObjectContext
        context.perform {
            keys.forEach { (key) in
                guard let chassis = data[key] as? [AnyHashable: Any] else { return }
                guard let module_id = chassis[WGJsonFields.module_id] as? String else { return }
                let predicate = NSPredicate(format: "%K == %@", #keyPath(Tankchassis.module_id), module_id)
                guard let newObject = NSManagedObject.findOrCreateObject(forClass:Tankchassis.self, predicate: predicate, context: context) as? Tankchassis else { return }

                newObject.mapping(fromJSON: chassis, into: context, completion: nestedRequestsCallback)
            }
            
            if context.hasChanges {
                do {
                    try context.save()
                } catch { 
                    print("\(#file) \(#function) at \(error.localizedDescription)")
                }
            }
        }
    }
}
